ipa-core/src/query/processor.rs

```diff
@@ -349,17 +362,111 @@ impl Processor {
         Some(status)
     }
 
-    /// Returns the query status.
+    /// This helper function is used to transform a [`BoxError`] into a
+    /// [`QueryStatusError::DifferentStatus`] and retrieve it's internal state. Returns [`None`]
+    /// if not possible.
+    fn downcast_state_error(be: BoxError) -> Option<QueryStatus> {
+        let ae = be.downcast::<crate::helpers::ApiError>().ok()?;
+        if let crate::helpers::ApiError::QueryStatus(QueryStatusError::DifferentStatus {
+            my_status,
+            ..
+        }) = *ae
+        {
+            return Some(my_status);
+        }
+        None
+    }
+
+    /// This helper is used by the in-memory stack to obtain the state of other shards via a
+    /// [`QueryStatusError::DifferentStatus`] error.
+    /// TODO: Ideally broadcast should return a value, that we could use to parse the state instead
+    /// of relying on errors.
+    #[cfg(feature = "in-memory-infra")]
+    fn get_state_from_error(
```

Comment:
there are a few issues here
- This logic really belongs to `BroadcastError`. 
- what happens if we compile this with real-world-infra flag?

Reply:
1. I don't think this belongs to BroadcastError. In fact, none of these new functions (`get_state_from_error` and `downcast_state_error`) have to deal with that type at all. Only `query_status` on broadcast.

2. Having 2 implementations under the same name won't compile. This is the same as `gateway::MpcTransportImpl`.
Also `InMemoryTransportError` won't be an available symbol in real-world-infra.

Reply:
the semantic of error downcasting does not seem to be query processor job. Likely we would need to do more for real world

> Having 2 implementations under the same name won't compile. This is the same as gateway::MpcTransportImpl.

is it going to be an issue when we implement this for HTTP?

Reply:
Don't think I'm following, let's chat about this one.

The http implementation for the downcast is here. As long as we use AplicationError this should handle it. We can change it to be something more specific as well.